### PR TITLE
xplat: sync to latest and fix bugs

### PR DESCRIPTION
**Pull request by jianchun**
Sync to latest `chakracore-master` and `chakracore`.

Interesting changes in the last 2 commits:
- Expose chakracore version.
- Use experimental String APIs to fix String encoding conversion bugs.

### CODE REVIEW

File: deps/chakrashim/core/lib/Jsrt/ChakraCommon.h [hunk 1]
```diff
@@ -2646,6 +2646,49 @@ typedef UINT32 DWORD;
         JsStringFree(
             _In_ char* stringValue);
 
+    // -------- Experimental new String APIs ----------------------------------
```
Comment (kunalspathak) [inline deps/chakrashim/core/lib/Jsrt/ChakraCommon.h:2649]: So are these APIs already in chakracore's master branch?

Comment (jianchun) [inline deps/chakrashim/core/lib/Jsrt/ChakraCommon.h:2649]: No. I'll push to a temp branch for the time being as these are incubating APIs.

File: deps/chakrashim/core/lib/Jsrt/Jsrt.cpp [hunk 2]
```diff
@@ -3433,6 +3433,178 @@ CHAKRA_API JsStringFree(_In_ char* stringValue)
...
+// -------- Experimental new String APIs ----------------------------------
+
+
+template <class SrcChar, class DstChar>
+static void CastCopy(const SrcChar* src, DstChar* dst, size_t count)
+{
+    const SrcChar* end = src + count;
+    while (src < end)
+    {
+        *dst++ = static_cast<DstChar>(*src++);
```
Comment (kunalspathak) [inline deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:3446]: Curious, what happens when `sizeof(SrcChar) > sizeof(DstChar)`? I see below that you are using this method where src is char16 and dst is char. Don't we miss out 8 bits in casting?

File: deps/chakrashim/core/lib/Jsrt/Jsrt.cpp [hunk 3]
```diff
@@ -3433,6 +3433,178 @@ CHAKRA_API JsStringFree(_In_ char* stringValue)
...
+    char* buffer,
+    size_t bufferSize,
+    _Out_opt_ size_t* length)
+{
+    const char16* str = nullptr;
+    size_t strLength = 0;
+    JsErrorCode errorCode = JsStringToPointer(value, &str, &strLength);
+    if (errorCode != JsNoError)
+    {
+        return errorCode;
```
Comment (kunalspathak) [inline deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:3513]: can you correct the indentation and run lint?

File: deps/chakrashim/core/lib/Jsrt/Jsrt.cpp (hunk 2 above)
Comment (jianchun) [inline deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:3446]: This is to support v8 String's "OneByte" (latin1) API. Yes, in that case the higher 8 bits are discarded (by design). If the chars are really latin1 this is not a problem. If the chars are not, reading OneByte result is wrong anyway.

File: deps/chakrashim/core/lib/Jsrt/Jsrt.cpp [hunk 4]
```diff
@@ -3433,6 +3433,178 @@ CHAKRA_API JsStringFree(_In_ char* stringValue)
...
+        if (length)
+        {
+            *length = count;
+        }
+    }
+
+    return JsNoError;
+}
+
+CHAKRA_API JsWriteStringUtf8(
```
Comment (kunalspathak) [inline deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:3573]: Can we rename methods `JsWrite*` to something like `JsWrite*ToBuffer`?

File: deps/chakrashim/core/lib/Jsrt/Jsrt.cpp (hunk 2 above)
Comment (kunalspathak) [inline deps/chakrashim/core/lib/Jsrt/Jsrt.cpp]: > If the chars are not, reading OneByte result is wrong anyway

So are you saying that `JsWriteString` should always be called for latin1 characters?

File: deps/chakrashim/core/lib/Jsrt/Jsrt.cpp (hunk 4 above)
Comment (jianchun) [inline deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:3573]: These names are unstable yet, could change in future. Feeling adding that is more clear but too long.

File: deps/chakrashim/core/lib/Jsrt/Jsrt.cpp (hunk 2 above)
Comment (kunalspathak) [inline deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:3446]: > If the chars are not, reading OneByte result is wrong anyway.

So are you saying `JavascriptString` passed to `JsWriteString` should always be latin1?

Comment (jianchun) [inline deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:3446]: It is under developers' control, and the behavior would be by design. Yes, supposedly it is only meaningful if the String is latin1 string.

File: deps/chakrashim/chakracore.gyp [hunk 5]
```diff
@@ -114,6 +115,8 @@
         'library_dirs': [ '<(PRODUCT_DIR)' ],
         'conditions': [
           ['OS=="win"', {
+            'libraries': [
```
Comment (kunalspathak) [inline deps/chakrashim/chakracore.gyp]: why is this needed?

Comment (jianchun) [inline deps/chakrashim/chakracore.gyp]: I forgot...

File: deps/chakrashim/src/v8v8.cc [hunk 6]
```diff
@@ -18,11 +18,15 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 // IN THE SOFTWARE.
 
+#include <algorithm>
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8v8.cc:21]: nit: don't need this i suppose?

File: deps/chakrashim/src/v8string.cc [hunk 7]
```diff
@@ -82,98 +82,57 @@ int String::Utf8Length() const {
...
-  if (length == 0) {
-    // bail out if we are required to write no chars
-    return 0;
-  }
-
-  jsrt::StringUtf8 str;
-  if (str.From(ref) != JsNoError) {
-    // error
-    return 0;
+int String::Write(uint16_t *buffer, int start, int length, int options) const {
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8string.cc:85]: does the API says to write into buffer starting from offset `start` and `length` characters?

File: deps/chakrashim/src/v8string.cc [hunk 8]
```diff
@@ -82,98 +82,57 @@ int String::Utf8Length() const {
...
   }
 
-  jsrt::StringConvert::CopyRaw(str + start, count, buffer, length);
-
-  if (count < length && !(options & String::NO_NULL_TERMINATION)) {
-    // include the null terminate
-    buffer[count++] = L'\0';
+  size_t count = 0;
+  if (JsWriteString((JsValueRef)this,
+    (char*)buffer + start, length, &count) == JsNoError) {
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8string.cc]: nit: don't need casting  of `(char*)` to buffer + start, right?

File: deps/chakrashim/src/v8v8.cc (hunk 6 above)
Comment (jianchun) [inline deps/chakrashim/src/v8v8.cc:21]: It is needed by `std::remove` used in this file. I moved it to front (recall cpplint complains about similar styles).

File: deps/chakrashim/src/v8string.cc (hunk 8 above)
Comment (kunalspathak) [inline deps/chakrashim/src/v8string.cc]: nvm, re-reading `JsWriteString` API. Can you atleast wrap it? `(char*)(buffer + start)` ?

File: deps/chakrashim/src/v8string.cc (hunk 7 above)
Comment (jianchun) [inline deps/chakrashim/src/v8string.cc:85]: https://github.com/nodejs/node-chakracore/blob/chakracore-master/deps/v8/include/v8.h#L2157

```
   * \param start The starting position within the string at which
   * copying begins.
   * \param length The number of characters to copy from the string.
```

**Comment by kunalspathak** [comment]
LGTM.

File: deps/chakrashim/src/v8string.cc (hunk 7 above)
Comment (jianchun) [inline deps/chakrashim/src/v8string.cc:85]: Hmm, I am unsure what it does exactly now. Will do some tests to find out later. (Maybe create a bug on this.)

Comment (jianchun) [inline deps/chakrashim/src/v8string.cc:85]: Reread old code, should be start in source string. I'll see if I can fix this in time, or create a future bug.

Good catch, thanks!
